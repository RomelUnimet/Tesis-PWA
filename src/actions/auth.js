import { fetchNoToken, fetchWithToken } from '../helpers/fetch'
import { types } from '../types/types'
//Lo haremos sin Axios para tener mas control sobre las peticiones

import Localbase from 'localbase';
import { decodeToken } from '../helpers/decodeToken';


const db = new Localbase('pwa-card-diary');


export const startLogin = ( email, password) => {

    return async ( dispatch ) => {
        
        const resp = await fetchNoToken('auth', { email, password }, 'POST');
        const body = await resp.json();

        if( body.ok){


            await db.collection('token').set([{
                token: body.token,
              }]);

            const info = await fetchWithToken('auth/info');
            
            const userInfo = await info.json();

            console.log(userInfo)

            await db.collection('settings').set(
                userInfo.settings,
              );
             
            await db.collection('cards').set(
                userInfo.cards
            );

            dispatch( login ({
                uid:body.uid,
            }))
            
            
        } else {
            console.log('Mostrar un mensaje de error')
            //Aqui debemos mostrar un mensaje de error
        }
    }  
}

export const startRegister = ( email, password) => {

    return async ( dispatch ) => {

        const resp = await fetchNoToken('auth/new', { email, password }, 'POST');
        const body = await resp.json();

        
        
        if( body.ok ){

            await db.collection('token').set([{
                token: body.token,
              }]);
            
            const settings = []
            settings.push(body.settings)

            await db.collection('settings').set(
                settings
              );

            await db.collection('cards').delete();
            
            for (let index = 0; index < body.cards.length; index++) {
                
                await db.collection('cards').add(
                    body.cards[index]
                );
    
            }

            dispatch( login ({
                uid:body.uid,
            }))

        } else {
            console.log('Mostrar un mensaje de error')
            //Aqui debemos mostrar un mensaje de error
        }
    }  
}




export const startChecking = () => {

    return async ( dispatch ) => {

        const isLogged = await db.collection('token').get();

        //Aca se muestra como si se hiciera 2 veces pero es mentira
        //Es donde se crean las colecciones en IndexedDB si no se tienen todavia
        await db.collection('cards').get();
        await db.collection('settings').get();
        await db.collection('tags').get();
        await db.collection('locations').get();

        if(isLogged.length === 0){

            const resp = await fetchWithToken('auth/renew');
            const body = await resp.json();

            if( body.ok){

                await db.collection('token').set([{
                                token: body.token,
                            }]);

                dispatch( login ({
                    uid:body.uid,
                }))
            } else {
                console.log('Body.ok dio false')
                //Aqui debemos mostrar un mensaje de error

                dispatch ( checkingFinish() )
            }

        } else {

            const token = isLogged[0].token;

            const payload = decodeToken(token);

            dispatch ( chekingToken({uid: payload.uid}) )

        }

        
    } 

}

const checkingFinish = () => {
    return { type: types.authChekingFinish }
}

const chekingToken = ( uid ) => {

    return { 
        type: types.authChekingToken,
        payload:  uid
    }
}

const login = ( user ) => {
    return {
        type: types.authLogin,
        payload: user        
    }
}




