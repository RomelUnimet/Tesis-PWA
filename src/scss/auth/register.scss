@import '../variables';


.register-container{
    display: flex;
    flex-direction: column;
    align-items: center;
    
}
.register-icon {
    fill: $accent-blue;
    padding: 0.8rem 1rem;
    border-radius: 1.2rem;
    box-shadow: 0px 0px 5px rgba(0, 0, 0, 0.25);
    background-color: white;
    position: relative;
    top: 40px;
    height: 3.5rem;
    width: 3.5rem;
    
}
.register-title{
    font-size: 1.6rem;
    margin-top: 6.5rem;
    font-weight: 700;
}

.form-container {
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    width: 85%;
    height: 25rem;
    background: #FFFFFF;
    box-shadow: 0px 0px 10px rgba(158, 110, 110, 0.25);
    border-radius: 1rem;
}
.register-input {

    width: 70%;
    border-width: 0 0 2px;
    border-color: $grey-color;
    text-align: center;
    height: 2rem;
    margin-top: 1rem;
    margin-bottom: 2rem;
    font-size: 1rem;
    font-weight: 500;
    color: #575757;

    &:focus{
        outline:none;
        color: $black-color;
        
    }
}

.btn-register {
    background-color: $accent-blue;
    border: none;
    padding: 0.8rem 3rem;
    border-radius: 10rem;
    color: white;
    font-size: 1.1rem;
    margin-top: 2rem;
    
    position: relative;
    top: 2rem;
}
