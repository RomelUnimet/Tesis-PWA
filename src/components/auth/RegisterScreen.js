import React from 'react'
import { useDispatch } from 'react-redux'
import { startRegister } from '../../actions/auth';
//import { startLogin } from '../../actions/auth';
import { useForm } from '../../hooks/useForm'

import '../../scss/auth/register.scss'

export const RegisterScreen = () => {


    const dispatch = useDispatch();

    const [ formValues, handleRegisterInputChange ] = useForm({
        email:'',
        password: ''
    })

    const { email, password } = formValues;

    const handleRegister = ( e ) =>{
        e.preventDefault();

        //la cosa de swall esta cool

        //Login Validation
        //dispatch( startLogin(email, password)); 
        dispatch( startRegister( email, password )); 
    }

    return (

        <div className="register-container">

            <h1 className="register-title">Create Account</h1>

            <div className="icon-container">
                <svg width="32" height="42" viewBox="0 0 32 42" xmlns="http://www.w3.org/2000/svg"
                    className="register-icon"
                >
                    <path d="M30.5455 0H1.45455C0.65 0 0 0.670312 0 1.5V40.5C0 41.3297 0.65 42 1.45455 42H30.5455C31.35 42 32 41.3297 32 40.5V1.5C32 0.670312 31.35 0 30.5455 0ZM18.7273 3.375H23.0909V13.2141L20.9773 11.625L18.7273 13.2844V3.375ZM28.7273 38.625H3.27273V3.375H16V17.2922C16 17.4469 16.0455 17.6016 16.1364 17.7281C16.1915 17.8086 16.2617 17.8769 16.3426 17.9293C16.4236 17.9816 16.5138 18.0168 16.6081 18.033C16.7024 18.0491 16.7988 18.0458 16.8918 18.0232C16.9848 18.0007 17.0726 17.9593 17.15 17.9016L20.9591 15.0938L24.6591 17.8781C24.7818 17.9719 24.9318 18.0234 25.0864 18.0234C25.4864 18.0234 25.8136 17.6859 25.8136 17.2734V3.375H28.7227V38.625H28.7273Z" />
                </svg>
            </div>


            <form onSubmit={ handleRegister }
             className="form-container"
             >
                <input 

                    type="text"
                    className="register-input"
                    placeholder="Email"
                    name="email"
                    value={ email }
                    onChange={ handleRegisterInputChange }
                />

                <input

                    type="password"
                    className="register-input"
                    placeholder="Password"
                    name="password"
                    value={ password }
                    onChange={ handleRegisterInputChange }
                />

                <button 
                    type="submit"
                    onClick={handleRegister}
                    className="btn-register" 
                    
                > 
                        Sign up 
                </button>
            </form>

            
        </div>
    )
}

