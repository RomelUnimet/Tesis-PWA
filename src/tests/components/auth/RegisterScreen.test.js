import { mount, configure } from "enzyme"
import { Provider } from 'react-redux'
import { store } from '../../../store/store'
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { RegisterScreen } from "../../../components/auth/RegisterScreen";


configure({ adapter: new Adapter() })


const wrapper = mount(
    <Provider store={store}>
        <RegisterScreen/>
    </Provider>
)

describe('Pruebas Register Screen', () => { 

    test('should render the register component', async () => { 

        expect(wrapper).toMatchSnapshot()        
     })

 })