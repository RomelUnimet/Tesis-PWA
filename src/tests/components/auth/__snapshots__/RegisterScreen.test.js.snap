// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas Register Screen should render the register component 1`] = `ReactWrapper {}`;
